group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -544,6 +573,81 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
             .setTopics(topicResponses);
     }
 
+    /**
+     * Remove expired offsets for group.
+     *
+     * @param groupId The group id.
+     * @param records The list of records to populate with offset commit tombstone records.
+     * @param offsetsRetentionMs The offset retention in milliseconds.
+     *
+     * @return The group id if the group no longer has any offsets remaining, empty otherwise.
+     */
+    public Optional<String> cleanupExpiredOffsets(String groupId, List<Record> records, long offsetsRetentionMs) {
+        TimelineHashMap<String, TimelineHashMap<Integer, OffsetAndMetadata>> offsetsByTopic = offsetsByGroup.get(groupId);
+        if (offsetsByTopic == null) {
+            return Optional.of(groupId);
+        }
+        try {
+            Group group = groupMetadataManager.group(groupId);
+            ExpirationCondition expirationCondition = group.expirationCondition();
+            Set<TopicPartition> expiredPartitions = new HashSet<>();
+            long currentTimestamp = time.milliseconds();
+            AtomicBoolean hasAllOffsetsExpired = new AtomicBoolean(true);
+            offsetsByTopic.forEach((topic, partitions) -> {
+                if (!expirationCondition.subscribedTopics.contains(topic)) {
```

Comment:
I have also thought of using this but the semantics are a bit different. From `GenericGroup#isSubscribedToTopic`:
```
    /**
     * Returns true if the consumer group is actively subscribed to the topic. When the generic
     * group does not know, because the information is not available yet or because it has
     * failed to parse the Consumer Protocol, it returns true to be safe.
     *
     * @param topic The topic name.
     * @return whether the group is subscribed to the topic.
     */
    public boolean isSubscribedToTopic(String topic) {
        return subscribedTopics.map(topics -> topics.contains(topic))
            .orElse(true);
    }
```

Reply:
i updated the code so that we pass in a boolean to `isSubscribedToTopic` to specify whether we want an empty subscribedTopics to be considered subscribed or not.

Reply:
Hum... Please help me to better understand this one. My understanding is that defaulting to `true` when the subscribed topics is not defined causes an issue for the generic group not using the consumer protocol type. Did I get this right?

If my understanding is correct, I wonder if we could actually change the implementation of `GenericGroup#isSubscribedToTopic` to something like this:

```
    public boolean isSubscribedToTopic(String topic) {
        return subscribedTopics.map(topics -> topics.contains(topic))
            .orElse(usesConsumerGroupProtocol());
    }
```

The idea is to only be conservative if the consumer protocol type is used. Would something like this work?

Reply:
I think the main issue is that the existing behavior in GroupMetadata#removeExpiredOffsets only considers what topics a group is subscribed to if the group is using the consumer group protocol AND is Stable. If a group is in any other state, it acts as if the group is not subscribed to any topic when expiring offsets.

here's my concern with the above suggestion:

let's say we have an empty group that uses the consumer group protocol. subscribedTopics will be empty as there are no members (set in `computeSubscribedTopics`). This will return `true` from `isSubscribedToTopic`. This is not aligned with the existing behavior which says if a group is empty and has a protocol type, we return an empty collection so that the group is considered not subscribed to any topics during offset expiration.
```
      case Some(_) if is(Empty) =>
        // no consumer exists in the group =>
        ...
        getExpiredOffsets(
          commitRecordMetadataAndOffset => currentStateTimestamp
            .getOrElse(commitRecordMetadataAndOffset.offsetAndMetadata.commitTimestamp), <Empty set for subscribedTopics>
        )
```

Reply:
Hum.. I am not sure to follow. In this case, isn't `subscribedTopics` going to be set to an optional containing an empty list?

Reply:
I am also confused by the fact that you did the change. Did you do it because it is correct or by mistake?

Reply:
> In this case, isn't subscribedTopics going to be set to an optional containing an empty list?
ah.. i misread and thought subscribedTopics becomes an empty optional, not an empty set. So we would return `false` which is the correct behavior.

> Did you do it because it is correct or by mistake?
which change, to add another argument? I added that so we can rely on using `isSubscribedToTopic` instead of passing in the subscribed topics set.

i will update with your suggestion, it makes sense to me.